#### Explore AI Academy: Regression Project


<div style="font-size: 35px">
    <font color='Green'> <b>CO2 Carbon Emissions Data Analysis</b></font> 

![agri_image.png](attachment:agri_image.png)

<a id="cont"></a>

## Table of Contents
* <b>[1. Project Overview**](#chapter1)
    * [1.1 Introduction**](#section_1_1)
        * [1.1.1 Problem Statement**](#sub_section_1_1_1)
    * [1.2 Objectives**](#section_1_2)
* <b>[2. Importing Packages**](#chapter2)
* <b>[3. Loading Data**](#chapter3)
* <b>[4. Data Cleaning**](#chapter4)
* <b>[5. Exploratory Data Analysis (EDA)**](#chapter5)
* <b>[6. Feature Engineering**](#chapter6)</b>
* [7. Modeling](#chapter7)
* [8. Model Performance](#chapter8)
* <b>[9. Conclusion**](#chapter10)</b>
* [10. References](#chapter11)

## 1. Project Overview <a class="anchor" id="chapter1"></a>

### 1.1 Introduction <a class="anchor" id="section_1_1"></a>

> <b> Objective of the Project:</b> Clearly state what the project aims to achieve or what problem it intends to solve. This helps to immediately orient readers to the purpose of your analysis.

> <b> Data Source:</b> Mention where the data comes from, including specifics about the dataset such as the period it covers, who collected it, and any important features or variables it includes. This establishes the context and relevance of the data.

> <b> Importance of the Study:</b> Explain why this project is important. Discuss its relevance or impact in its field, whether it’s improving a business process, advancing scientific understanding, or influencing policy decisions.

> <b> Key Questions or Hypotheses:</b> Outline the main questions you aim to answer or hypotheses you plan to test. This sets clear goals for your analysis and what the audience can expect to learn.

> <b> Methodology Overview:</b> Provide a brief overview of the methods and techniques you plan to use in your analysis. This doesn’t need to be detailed here but should give a taste of the analytical approach and tools.

> <b> Structure of the Notebook:</b> Describe the layout of the notebook. Explain the sections or parts of the analysis to help readers navigate through the content.

> <b> Expected Outcomes:</b> If applicable, mention what kind of results or findings you anticipate. This could be specific insights, predictive model performance, or impact assessment.

- #### 1.1.1 Problem Statement <a class="anchor" id="sub_section_1_1_1"></a>

### 1.2 Objectives <a class="anchor" id="section_1_2"></a>

+ To apply exploratory data analysis.
+ To implement feature engineering techniques to extract meaningful information.
+ To model and assess various supervised machine learning algorithms for the prediction ....

## 2. Importing Packages <a class="anchor" id="chapter2"></a>



+ For data manipulation and analysis, `Pandas` and `Numpy`.
+ For data visualization, `Matplotlib` and `Seaborn`.
+ 

In [1]:
# Libraries for data loading, manipulation and analysis

import numpy as np
import pandas as pd
import csv
import seaborn as sns
import matplotlib.pyplot as plt

# Displays output inline
%matplotlib inline

# Libraries for Handing Errors
import warnings
warnings.filterwarnings('ignore')

## 3. Loading Data <a class="anchor" id="chapter3"></a>

The data used for this project was located in the 2401PTDS_REGRESSION_PROJECT folder on github. To better manipulate and analyse the co2_emissions_from_agri.csv file, it was loaded into a Pandas Data Frame using the Pandas function, `.read_csv()`. We defined the dataset using the df function and will be used throughout the notebook.

In [2]:
# loading dataset
df = pd.read_csv("co2_emissions_from_agri.csv")
df.head(10) 

Area  Year  Savanna fires  Forest fires  Crop Residues  \
0  Afghanistan  1990        14.7237        0.0557       205.6077   
1  Afghanistan  1991        14.7237        0.0557       209.4971   
2  Afghanistan  1992        14.7237        0.0557       196.5341   
3  Afghanistan  1993        14.7237        0.0557       230.8175   
4  Afghanistan  1994        14.7237        0.0557       242.0494   
5  Afghanistan  1995        14.7237        0.0557       243.8152   
6  Afghanistan  1996        38.9302        0.2014       249.0364   
7  Afghanistan  1997        30.9378        0.1193       276.2940   
8  Afghanistan  1998        64.1411        0.3263       287.4346   
9  Afghanistan  1999        46.1683        0.0895       247.4980   

   Rice Cultivation  Drained organic soils (CO2)  Pesticides Manufacturing  \
0            686.00                          0.0                 11.807483   
1            678.16                          0.0                 11.712073   
2            686.00                          0.0                 11.712073   
3            686.00                          0.0                 11.712073   
4            705.60                          0.0                 11.712073   
5            666.40                          0.0                 11.712073   
6            686.00                          0.0                 11.712073   
7            705.60                          0.0                 11.712073   
8            705.60                          0.0                 11.712073   
9            548.80                          0.0                 11.712073   

   Food Transport  Forestland  ...  Manure Management  Fires in organic soils  \
0         63.1152   -2388.803  ...           319.1763                     0.0   
1         61.2125   -2388.803  ...           342.3079                     0.0   
2         53.3170   -2388.803  ...           349.1224                     0.0   
3         54.3617   -2388.803  ...           352.2947                     0.0   
4         53.9874   -2388.803  ...           367.6784                     0.0   
5         54.6445   -2388.803  ...           397.5498                     0.0   
6         53.1637   -2388.803  ...           465.2050                     0.0   
7         52.0390   -2388.803  ...           511.5927                     0.0   
8         52.7050   -2388.803  ...           541.6598                     0.0   
9         35.7630   -2388.803  ...           611.0611                     0.0   

   Fires in humid tropical forests  On-farm energy use  Rural population  \
0                              0.0                 NaN         9655167.0   
1                              0.0                 NaN        10230490.0   
2                              0.0                 NaN        10995568.0   
3                              0.0                 NaN        11858090.0   
4                              0.0                 NaN        12690115.0   
5                              0.0                 NaN        13401971.0   
6                              0.0                 NaN        13952791.0   
7                              0.0                 NaN        14373573.0   
8                              0.0                 NaN        14733655.0   
9                              0.0                 NaN        15137497.0   

   Urban population  Total Population - Male  Total Population - Female  \
0         2593947.0                5348387.0                  5346409.0   
1         2763167.0                5372959.0                  5372208.0   
2         2985663.0                6028494.0                  6028939.0   
3         3237009.0                7003641.0                  7000119.0   
4         3482604.0                7733458.0                  7722096.0   
5         3697570.0                8219467.0                  8199445.0   
6         3870093.0                8569175.0                  8537421.0   
7         4008032.0                8916862.0                  8871958.0  

To set the maximum number of columns to be displayed, the `pd.set_option()` function was put in place.

In [3]:
# displays unlimited number of columns
pd.set_option("display.max_columns", None)

<div class="alert alert-block alert-danger">
<b>To prevent any major unnecessary changes occurring to the original data</b> , a copy of the dataframe was made using the df.copy() method and referred to as `df_copy`.
</div>

In [4]:
# The copy of the dataframe
df_copy = df.copy()

First look at the shape of the dataframe by using `.shape`.

In [5]:
# Displays the number of rows and columns
df_copy.shape

(6965, 31)

**Results** : The dataset consists of 6965 rows (observations) and 31 columns (features). 

One of the univariate analysis (non-graphical), is the method `.info()`. This method outputs important details about the dataset which includes the columns: the data types **(Dtype)** of the columns and the count of non-null values.

In [6]:
# Display summary information about the DataFrame.
df_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6965 entries, 0 to 6964
Data columns (total 31 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Area                             6965 non-null   object 
 1   Year                             6965 non-null   int64  
 2   Savanna fires                    6934 non-null   float64
 3   Forest fires                     6872 non-null   float64
 4   Crop Residues                    5576 non-null   float64
 5   Rice Cultivation                 6965 non-null   float64
 6   Drained organic soils (CO2)      6965 non-null   float64
 7   Pesticides Manufacturing         6965 non-null   float64
 8   Food Transport                   6965 non-null   float64
 9   Forestland                       6472 non-null   float64
 10  Net Forest conversion            6472 non-null   float64
 11  Food Household Consumption       6492 non-null   float64
 12  Food Retail         

**Results** : The dataset consists of the following datatypes: 29 float64, 1 integer64 and 1 object.

### Dataset Attributes

- Savanna fires: Emissions from fires in savanna ecosystems.
- Forest fires: Emissions from fires in forested areas.
- Crop Residues: Emissions from burning or decomposing leftover plant material after crop harvesting.
- Rice Cultivation: Emissions from methane released during rice cultivation.
- Drained organic soils (CO2): Emissions from carbon dioxide released when draining organic soils.
- Pesticides Manufacturing: Emissions from the production of pesticides.
- Food Transport: Emissions from transporting food products.
- Forestland: Land covered by forests.
- Net Forest conversion: Change in forest area due to deforestation and afforestation.
- Food Household Consumption: Emissions from food consumption at the household level.
- Food Retail: Emissions from the operation of retail establishments selling food.
- On-farm Electricity Use: Electricity consumption on farms.
- Food Packaging: Emissions from the production and disposal of food packaging materials.
- Agrifood Systems Waste Disposal: Emissions from waste disposal in the agrifood system.
- Food Processing: Emissions from processing food products.
- Fertilizers Manufacturing: Emissions from the production of fertilizers.
- IPPU: Emissions from industrial processes and product use.
- Manure applied to Soils: Emissions from applying animal manure to agricultural soils.
- Manure left on Pasture: Emissions from animal manure on pasture or grazing land.
- Manure Management: Emissions from managing and treating animal manure.
- Fires in organic soils: Emissions from fires in organic soils.
- Fires in humid tropical forests: Emissions from fires in humid tropical forests.
- On-farm energy use: Energy consumption on farms.
- Rural population: Number of people living in rural areas.
- Urban population: Number of people living in urban areas.
- Total Population - Male: Total number of male individuals in the population.
- Total Population - Female: Total number of female individuals in the population.
- total_emission: Total greenhouse gas emissions from various sources.
- Average Temperature °C: The average increasing of temperature (by year) in degrees Celsius,
 

CO2 is recorded in kilotonnes (kt): 1 kt represents 1000 kg of CO2.

The feature "Average Temperature C°", represents the average yearly temperature increase. For example, if it is 0.12, it means that the temperature in that specific location increased by 0.12 degrees Celsius.

Forestland is the only feature that exhibits negative emissions due to its role as a carbon sink. Through photosynthesis, forests absorb and store carbon dioxide, effectively removing it from the atmosphere. Sustainable forest management, along with afforestation and reforestation efforts, further contribute to negative emissions by increasing carbon sequestration capacity.



## 4. Data Cleaning <a class="anchor" id="chapter4"></a>

<div class="alert alert-block alert-info">
<b>Data cleaning</b> is a crucial step in the data analysis process, involving the correction or removal of incorrect, corrupted, duplicate, or incomplete data within a dataset. Through various techniques such as filling missing values, removing outliers, and standardizing data formats, it ensures the accuracy and reliability of subsequent analyses and decision-making.
</div>

The `check_null_values` function iterates over each column in a DataFrame to identify and print the count of null values, providing column-specific null counts.

In [7]:
def check_null_values(df_copy):
    """
    Print the count of null values for each column in a DataFrame.

    This function iterates through each column in the DataFrame to check for the presence of null values.
    If a column contains null values, it prints the column name along with the number of null values.

    Parameters:
    df (DataFrame): The pandas DataFrame to check for null values.

    Returns:
    None: This function does not return a value; it only prints information.
    """
    for column in df_copy:
        if df_copy[column].isnull().any():
            print('{0} has {1} null values'.format(column, df_copy[column].isnull().sum()))

In [8]:
check_null_values(df_copy)

Savanna fires has 31 null values
Forest fires has 93 null values
Crop Residues has 1389 null values
Forestland has 493 null values
Net Forest conversion has 493 null values
Food Household Consumption has 473 null values
IPPU has 743 null values
Manure applied to Soils has 928 null values
Manure Management has 928 null values
Fires in humid tropical forests has 155 null values
On-farm energy use has 956 null values


**Results**:

The `count_duplicate_rows` function checks a DataFrame for duplicate entries by utilizing the duplicated method, which identifies duplicates, and then summarizes the total number of duplicate rows. This function is useful for data cleaning processes, ensuring uniqueness in datasets by returning the count of duplicate rows.

In [9]:
def count_duplicate_rows(df_copy):
    """
    Count the number of duplicate rows in a DataFrame.

    This function calculates the total number of duplicate rows in the DataFrame by calling the `duplicated` method,
    which marks duplicates as `True`, and then sums these cases.

    Parameters:
    df_copy (pandas.DataFrame): The DataFrame to check for duplicates.

    Returns:
    int: The count of duplicate rows.
    """
    duplicate_count = df_copy.duplicated().sum()
    return duplicate_count


In [10]:
count_duplicate_rows(df_copy)

0

**Results**: There are no duplicate rows in the dataset 

The `preprocess_data` function performs essential data cleaning operations on a DataFrame, setting missing values, filling in gaps with statistical measures (mean and median), and dropping non-essential columns. After processing, it prints the first two rows of the cleaned DataFrame for quick inspection and returns the entire modified DataFrame for further use.


In [11]:
def preprocess_data(df_copy):
    """
    Perform basic data cleaning on the DataFrame.

    This includes:
    - Setting missing 'Crop Residues' to 0.
    - Filling missing Forestland and Net Forest conversion with the mean and median, respectively.
    - Filling missing Savanna fires and Forest fires with the median and mean, respectively.
    - Dropping columns with many null values or that are not required.


    Returns:
        DataFrame: The cleaned DataFrame.
    """
    
    # Assign zero Crop Residues to data where the number of Crop Residues is missing
    df_copy['Crop Residues'] = df_copy['Crop Residues'].fillna(0)
    
    # Fill missing Forestland with the mean score
    df_copy['Forestland'] = df_copy['Forestland'].fillna(df_copy['Forestland'].mean())
    
    # Fill missing Net Forest conversion with the median number of Net Forest Conversions
    df_copy['Net Forest conversion'] = df_copy['Net Forest conversion'].fillna(df_copy['Net Forest conversion'].median())
    
    # Fill missing Forest fires scores with the median fires
    df_copy['Forest fires'] = df_copy['Forest fires'].fillna(df_copy['Forest fires'].median())
    

In [12]:
# look at the top 2
preprocess_data(df_copy.head(2))

#### <a href=#cont>Back to Table of Contents</a>

---

## 5. Exploratory Data Analysis (EDA) <a class="anchor" id="chapter5"></a>


To give a better understanding of the variables and the relationships between them, we set out to do an **Exploratory Data Analysis (EDA)** of our dataset. The main tasks includes investigating and summarizing the dataframe's main characteristics by data visualization methods and statistical analyses. Furthermore, investigating the dataset’s key features, summarizing its central characteristics, and employing both data visualisation techniques and statistical analyses to draw meaningful insights that can guide further research and data-driven decision making.
 

#### Filter the DataFrame 

In [13]:
df.describe().T

count          mean           std  \
Year                             6965.0  2.005125e+03  8.894665e+00   
Savanna fires                    6934.0  1.188391e+03  5.246288e+03   
Forest fires                     6872.0  9.193022e+02  3.720079e+03   
Crop Residues                    5576.0  9.987063e+02  3.700345e+03   
Rice Cultivation                 6965.0  4.259667e+03  1.761383e+04   
Drained organic soils (CO2)      6965.0  3.503229e+03  1.586145e+04   
Pesticides Manufacturing         6965.0  3.334184e+02  1.429159e+03   
Food Transport                   6965.0  1.939582e+03  5.616749e+03   
Forestland                       6472.0 -1.782829e+04  8.183221e+04   
Net Forest conversion            6472.0  1.760564e+04  1.011575e+05   
Food Household Consumption       6492.0  4.847580e+03  2.578914e+04   
Food Retail                      6965.0  2.043211e+03  8.494249e+03   
On-farm Electricity Use          6965.0  1.626681e+03  9.343182e+03   
Food Packaging                   6965.0  1.658630e+03  1.148134e+04   
Agrifood Systems Waste Disposal  6965.0  6.018445e+03  2.215674e+04   
Food Processing                  6965.0  3.872724e+03  1.983822e+04   
Fertilizers Manufacturing        6965.0  3.035723e+03  1.169303e+04   
IPPU                             6222.0  1.999150e+04  1.114209e+05   
Manure applied to Soils          6037.0  9.232256e+02  3.226992e+03   
Manure left on Pasture           6965.0  3.518027e+03  9.103556e+03   
Manure Management                6037.0  2.263345e+03  7.980542e+03   
Fires in organic soils           6965.0  1.210316e+03  2.266985e+04   
Fires in humid tropical forests  6810.0  6.684529e+02  3.264879e+03   
On-farm energy use               6009.0  3.008982e+03  1.263786e+04   
Rural population                 6965.0  1.785774e+07  8.901521e+07   
Urban population                 6965.0  1.693230e+07  6.574362e+07   
Total Population - Male          6965.0  1.761963e+07  7.603993e+07   
Total Population - Female        6965.0  1.732447e+07  7.251711e+07   
total_emission                   6965.0  6.409124e+04  2.283130e+05   
Average Temperature °C           6965.0  8.729891e-01  5.559295e-01   

                                           min            25%           50%  \
Year                               1990.000000    1997.000000  2.005000e+03   
Savanna fires                         0.000000       0.000000  1.651850e+00   
Forest fires                          0.000000       0.000000  5.179000e-01   
Crop Residues                         0.000200      11.006525  1.036982e+02   
Rice Cultivation                      0.000000     181.260800  5.348174e+02   
Drained organic soils (CO2)           0.000000       0.000000  0.000000e+00   
Pesticides Manufacturing              0.000000       6.000000  1.300000e+01   
Food Transport                        0.000100      27.958600  2.049628e+02   
Forestland                      -797183.079000   -2848.350000 -6.292000e+01   
Net Forest conversion                 0.000000       0.000000  4.444000e+01   
Food Household Consumption            0.000000      11.399950  1.554711e+02   
Food Retail                           0.000000      26.818500  1.720426e+02   
On-farm Electricity Use               0.000000       8.037600  2.912070e+01   
Food Packaging                        0.000000      67.631366  7.401813e+01   
Agrifood Systems Waste Disposal       0.340000      86.680500  9.012757e+02   
Food Processing                       0.000100     209.587728  3.447602e+02   
Fertilizers Manufacturing             0.001900     360.358799  1.115052e+03   
IPPU                                  0.000000      39.031525  8.037066e+02   
Manure applied to Soils               0.049000      16.303000  1.204439e+02   
Manure left on Pasture                0.000700     139.669900  9.725674e+02   
Manure Management                     0.432900      37.632100  2.698563e+02   
Fires in organic soils                0.000000       0.000000  0.000000e+00   
Fires in humid 

In [19]:
# Identify outliers based on IQR in 'total_emission' column
Q1 = df['total_emission'].quantile(0.25)
Q3 = df['total_emission'].quantile(0.75)
IQR = Q3 - Q1
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR
outliers_price = (df['total_emission'] < lower_bound) | (df['total_emission'] > upper_bound)

# Print the number of outliers
print(f"Number of outliers in total_emission: {outliers_price.sum()}")

Number of outliers in total_emission: 506


*Result*: There are a total of 506 outliers in the total_emissions column. An outlier is a data point that differs significantly from other observations. An outlier may be due to a variability in the measurement, an indication of novel data, or it may be the result of experimental error; the latter are sometimes excluded from the data set.

In [15]:
# Remove rows with outliers
df = df[~outliers_price]

Calculate the number of each type of entry in the dataset, distinguishing between categories such as movies and TV shows to understand the overall composition and distribution.

In [23]:
# Get the count of entries per area in the dataset
Area_count = df_copy['Area'].value_counts()

# Display the top 5 most common Areas
Area_count.head()

Area
Afghanistan    31
Isle of Man    31
Montserrat     31
Morocco        31
Mozambique     31
Name: count, dtype: int64